src/service_control_client_impl.cc

```diff
@@ -188,6 +243,177 @@ Status ServiceControlClientImpl::Check(const CheckRequest& check_request,
+      // Specified API Key has expired.
+      return Status(Code::INVALID_ARGUMENT,
+                    "API key expired. Please renew the API key.");
+
+    case ::google::api::servicecontrol::v1::QuotaError::
+        PROJECT_STATUS_UNVAILABLE:
+    // The backend server for looking up project id/number is unavailable.
+    case ::google::api::servicecontrol::v1::QuotaError::
+        SERVICE_STATUS_UNAVAILABLE:
+    // The backend server for checking service status is unavailable.
+    case ::google::api::servicecontrol::v1::QuotaError::
+        BILLING_STATUS_UNAVAILABLE:
+      // The backend server for checking billing status is unavailable.
+      // Fail open for internal server errors per recommendation
+      return Status::OK;
+
+    default:
+      return Status(
+          Code::INTERNAL,
+          std::string("Request blocked due to unsupported error code: ") +
+              std::to_string(error.code()));
+  }
+
+  return Status::OK;
+}
+
+void ServiceControlClientImpl::Quota(
+    const ::google::api::servicecontrol::v1::AllocateQuotaRequest&
+        quota_request,
+    ::google::api::servicecontrol::v1::AllocateQuotaResponse* quota_response,
+    DoneCallback on_quota_done, TransportQuotaFunc check_transport) {
+  ++total_called_quotas_;
+  if (check_transport == NULL) {
+    on_quota_done(Status(Code::INVALID_ARGUMENT, "transport is NULL."));
+    return;
+  }
+
+  Status status = quota_aggregator_->Quota(quota_request, quota_response);
+  if (status.error_code() == Code::NOT_FOUND) {
+    // Makes a copy of check_request so that on_done() callback can use
+    // it to call CacheResponse.
+    ::google::api::servicecontrol::v1::AllocateQuotaRequest*
+        quota_request_copy =
+            new ::google::api::servicecontrol::v1::AllocateQuotaRequest(
+                quota_request);
+
+    std::shared_ptr<QuotaAggregator> quota_aggregator_copy = quota_aggregator_;
+    check_transport(*quota_request_copy, quota_response,
+                    [this, quota_aggregator_copy, quota_request_copy,
+                     quota_response, on_quota_done](Status status) {
+
+                      if (status.ok()) {
+                        quota_aggregator_copy->CacheResponse(
+                            *quota_request_copy, *quota_response);
+                      } else {
+                        GOOGLE_LOG(ERROR) << "Failed in Check call: "
+                                          << status.error_message();
+                      }
+
+                      delete quota_request_copy;
+
+                      on_quota_done(convertResponseStatus(*quota_response));
+                    });
+
+    ++send_quotas_in_flight_;
+    return;
+
+  } else if (status.error_code() == Code::INVALID_ARGUMENT ||
+             status.error_code() == Code::UNAVAILABLE) {
+    on_quota_done(status);
+  } else {
+    // Status::OK, return response status from AllocateQuotaResponse
+    on_quota_done(convertResponseStatus(*quota_response));
```

Comment:
Coverting Response to Status is better done in api_manager code.  This library is just for cache and aggregation.